myapp/utility.py

```diff
@@ -1,116 +1,880 @@
-import schema
+from functools import wraps
+from flask import make_response, render_template, jsonify, redirect, url_for
+from flask.ext.login import current_user
+from myapp import models as m
+from myapp import (app, db)
+import linecache
+import sys
+import re
+ds = db.session
 
 
-# Utility functions
-# Generate a lab's id according to lab_name, class_name and prof_name
-def Generate_lab_id(lab_name,class_name,prof_name):
+# --- Debug/Logging functions ---
+# make things easier to spot in terminal by printing them out
+def cprint(*args):
+    for s in args:
+        print('\n\n\n')
+        print(s)
+        print('\n\n\n')
+
+
+# Check the existence of args in a form
+# and generate an error message accordingly
+def check_existence(form, *args):
+    err_msg = ''
+    for a in args:
+        if not(a in form):
+            err_msg += (a+' is not defined'+'\n')
+    if err_msg != '':
+        app.logger.error(err_msg)
+    return err_msg
+
+
+# generate the error message by receiving error info
+def generate_err_msg(name1, value1, name2, value2):
+    return '{0}:{1} and {2}:{3} are different'.format(name1, value1, name2,
+                                                      value2)
+
+
+# --- ID generators/decomposers ---
+def generate_lab_id(lab_name, class_name, prof_name):
     return lab_name+'_'+class_name+'_'+prof_name
 
 
-# Generate a row's id according to lab_id, row_name
-def Generate_row_id(lab_id,row_name):
-    return lab_id+':'+row_name
+def generate_experiment_id(lab_id, experiment_name):
+    return lab_id+':'+experiment_name
 
 
-# Generate a data's id according to row_id, student_name
-def Generate_data_id(row_id,student_name):
-    return row_id+':'+student_name
+def generate_observation_id(experiment_id, student_name):
+    return experiment_id+':'+student_name
+
+
+# The user_id is the same as username at this point,
+# both of them must be unique
+def generate_user_id(username):
+    return username
+
+
+def generate_class_id(class_name, class_time):
+    return class_name+'_'+class_time
+
+
+def decompose_lab_id(lab_id):
+    return {'lab_name': lab_id.split('_')[0],
+            'prof_name': lab_id.split('_')[1],
+            'class_name': lab_id.split('_')[2]}
+
+
+# --- Check existence ---
+def lab_exists(lab_id):
+    return ds.query(m.Lab).filter(m.Lab.id == lab_id).count() > 0
+
+
+def experiment_exists(experiment_id):
+    return ds.query(m.Experiment).filter(m.Experiment.id == experiment_id).count() > 0
+
+
+def observation_exists(observation_id):
+    return ds.query(m.Observation).filter(m.Observation.id == observation_id).count() > 0
+
+
+def class_exists(class_id):
+    return ds.query(m.Class).filter(m.Class.id == class_id).count() > 0
+
+
+def user_exists(user_id):
+    return ds.query(m.User).filter(m.User.id == user_id).count() > 0
+
+
+# --- Query data ---
+# Get the Lab object from the database by lab_id
+def query_lab(lab_id):
+    return ds.query(m.Lab).filter(m.Lab.id == lab_id).first()
```

Comment:
This is another method called one in the document so i decide to use it instead of first.